### PR TITLE
[Component] Integrate DS Select + Multiselect

### LINKED ISSUE #197: Spike: Bring the DS Single and Multiselect into Storybook
Explore implementing DSR Select components by matching the Design System's HTML layout instead of attempting to replicate the styling using the `react-select` library. 
 
## Variants
- [Single select](https://cfpb.github.io/design-system/components/selects#single-select)
- [Multi-select](https://cfpb.github.io/design-system/components/selects#multiselect)

## We want to understand  
- What are the Pros / Cons of using primarily HTML instead of the React library?
- Any blockers?

### PR DESCRIPTION
**Pull request by meissadia**
Phase 1 for #197 

## Changes

- Adds DS Select
- Adds DS Multiselect.
- Pending issues outlined [here](https://github.com/cfpb/design-system-react/issues/197#issuecomment-1800430600)

## How to test this PR

`yarn vitest Select`

### Sidebar
Screenshot 2024-01-31 at 3 42 53 PM

### Custom Overview
Screenshot 2024-01-31 at 3 43 00 PM
 
### Single select Overview
Screenshot 2024-01-31 at 3 43 23 PM

### Multiselect Overview
Screenshot 2024-01-31 at 3 43 37 PM

### CODE REVIEW

**Comment by meissadia** [comment]
@shindigira @billhimmelsbach 

In talking with @natalia-fitzgerald I've learned there are other teams experimenting with the DSR so we'd like to merge the current progress we've made on integrating the DS Select/Multiselect into DSR with the understanding that we will pick up work on the pending issues once our [proposed changed to the DS Multiselect](https://github.com/cfpb/design-system/pull/1801) are merged and released.
 
Moving this PR into `ready for review`.

**Review by natalia-fitzgerald — APPROVED** [review]

**Comment by meissadia** [comment]
@shindigira 

> ### Second error
Good catch, this actually *is* a blocker to this component being useable, so we'll have to wait on merging these changes.  This requires our [DS PR](https://github.com/cfpb/design-system/pull/1801) to be merged, then for an update to the DS npm package to be released.  That said, this PR should still be approved if things look good to avoid delays once the other dependencies are ready, we will simply wait to merge. 

> 1. I think we should separate `Select` and `Multiselect` into two components. There's no overlap in markup or logic between the two.

I've added the export of SelectSingle and SelectMulti so folks have the ability to directly import/use either one.  I'm still a fan of having a consolidated Select for ease of use, with devs able to switch between functionalities with a single flag. 

> 2. Do we want to hardwire the `Tag` component or create a separate component and feed it back in as a prop?

We've been working toward [unifying the implementation of `Tag` across the CFPB DS](https://github.com/cfpb/design-system-react/issues/236), so yes I think we'd want SelectMulti to always use the the single, approved Tag implementation, as opposed to making it customizable.  

> 3. Don't have this PR close issue [Spike: Bring the DS Single and Multiselect into Storybook #197](https://github.com/cfpb/design-system-react/issues/197) yet.

👍🏾 

> 4. **One important thing I don't agree with is that the Select component is now an "uncontrolled component" which somewhat goes against the React idiosyncracies.**

There is nothing stopping implementors from passing a `value` prop and making SelectSingle a controlled component.  

For SelectMulti, since we're leveraging the DS Multiselect which has it's own highly customized DOM element generation and event processing, we simply don't have the ability to control it.

**Review by natalia-fitzgerald — APPROVED** [review]

**Comment by meissadia** [comment]
Our Multiselect changes have been released in DS v0.34.0.  We've upgraded the dependencies here, tested the Multiselect, everything there looks good. 

New blocker: Unit tests are failing on this branch with `TypeError: Unknown file extension ".svg" for ...`. 

Icons work fine in DSR so it seems isolated to the testing environment. 

- 🟢 Tried upgrading `main` branch to DS 0.34.0, tests run fine.  So the issue is introduced somewhere in this branch.
- 🟥 Tried rolling back DS version on this branch to 0.31.0, tests fail
- 🟥 Tried adding a jest.config.js to apply 'jest-svg-transformer', tests fail

I'll need to keep digging.

**Comment by meissadia** [comment]
Important to note that importing of SVGs works in DSR (tests work for DSR <Icon/>) but not when DS is the one importing an SVG.  Also this seems to be a `jest`/`vitest` issue since, as noted at the end of this update, I have stumbled upon a way to get DS elements with DS icons to render within Storybook.  The error is a blocker when running unit tests.

---

Things tried to fix the `.svg` issue
- 🟥 Expanding Jest configuration to use a `transformer` to handle SVGs: https://stackoverflow.com/a/59062195
- 🟥 Using `jest-svg-transformer`: https://stackoverflow.com/a/63042067
- 🟥 Using `jest-transform-stub` similar to `jest-svg-transformer`: https://stackoverflow.com/a/67585510
- 🟥 Tried using `vitest`s "optimizer" as mentioned in the associated github thread but that didn't solve the error: https://github.com/vitest-dev/vitest/issues/3862#issuecomment-1662483406

Existing config
- We already have `svgLoader` configured in `vite.config.ts`, but this must only apply to our DSR code, not other dependencies, and maybe not `vitest`?
      - Example of my attempts to add other transformer (svgo) - unsuccessful
        ```
        svgLoader({
          defaultImport: 'url',
          svgoConfig: {
            multipass: true
          }
        }),
        ```
---

One great find is that the following fixes the issue of DS icons not rendering in Vite!!!  So we could remove our implementation of DSR being responsible for rendering Multiselect's selected item tags. 

- vite.config.ts
```
svgLoader({ 
  defaultImport: 'raw'
}),
```

**Comment by meissadia** [comment]
Today tried:
- Upgrading `vitest` and `@vitest/coverage-istanbul` to latest version in order to be able to use the `deps.optimizer` config option which seems like it might solve our issue (I hadn't realized I needed this upgrade during yesterday's failure to implement this config).  
  - Those upgrades resulted in error message `expect is not defined` when trying to run a unit test via `vitest`. Currently debugging that.

**Comment by meissadia** [comment]
@billhimmelsbach Apologies, there's more component tests being affected than just the Multiselect
 
```
 FAIL  src/components/Select/Select.test.tsx [ src/components/Select/Select.test.tsx ]
 FAIL  src/components/Table/Table.test.tsx [ src/components/Table/Table.test.tsx ]
 FAIL  src/components/Alert/Alert.test.tsx [ src/components/Alert/Alert.test.tsx ]
 FAIL  src/components/Link/Link.test.tsx [ src/components/Link/Link.test.tsx ]
 FAIL  src/components/TextInput/TextInput.test.tsx [ src/components/TextInput/TextInput.test.tsx ]
 FAIL  src/components/Well/Well.test.tsx [ src/components/Well/Well.test.tsx ]
 FAIL  src/components/TextIntroduction/TextIntroduction.test.tsx [ src/components/TextIntroduction/TextIntroduction.test.tsx ]
TypeError: Unknown file extension ".svg" for /Users/diam/cfpb/design-system-react/node_modules/@cfpb/cfpb-icons/src/icons/error.svg
```

**Comment by billhimmelsbach** [comment]
No worries @meissadia, maybe try copying in your work on the authenticated landing page into the sbl-frontend?

**Comment by meissadia** [comment]
I've got all tests passing but have introduced some major lag in the generation of test coverage reporting
 
## How we got tests passing

### Applied some upgrades
- Upgrade vitest to 1.2.1
- Upgrade @vitejs/plugin-react to 4.2.1
- Upgrade @vitest/coverage-istanbul to 1.2.1

These upgrades got us passed the `expect is not defined` error mentioned above, but introduced an issue with importing svg files
error-svg-unknown

### Updated `vite.config.ts`
Per the recommendation of the Vitest developer, I've been exploring the `test.server.deps.inline` configuration
(https://github.com/vitest-dev/vitest/issues/3862#issuecomment-1662483406) and found that the following eliminated our SVG error.
```
server: {
  deps: {
    inline: ['@cfpb']
  }
}
```

Unfortunately, this introduces a significant delay (~2 minutes) in the generation of our code coverage reports due to a large quantity of seemingly unrelated files (i.e. from `loadash`, `@storybook`, many others) being included in the coverage report, even when testing a single file where the actual tests complete in under 4sec (ex: `yarn test Alert`).
 
## What I've tried to address extraneous coverage reporting
### Trying to exclude node_modules from coverage
I have searched for ways to exclude these unrelated files from the coverage reports.  Vitest docs indicate that `node_modules` should be excluded by default
- https://vitest.dev/config/#coverage-exclude
- https://vitest.dev/config/#exclude

I have tried to explicitly exclude `node_modules` from the `exclude`, `coverage.exclude`, and `test.exclude` configuration objects, but none have had any effect on the extraneous coverage reporting. 

```
import { configDefaults } from 'vitest/config';
export default defineConfig(() => ({
...
  exclude: [...configDefaults.exclude, '**/node_modules/**'],
  coverage: {
    exclude: [...configDefaults.coverage.exclude, '**/node_modules/**']
  },
  test: {
    exclude: [...configDefaults.exclude, '/node_modules/*/*'],
    ...
  },
...
}));
```

## Bright side: Things work well on subsequent test runs...
Running `yarn test` will start `vitest` in watch mode.  After the initial run/coverage, if I simply hit `Enter` to trigger another test run, coverage reporting is fast (under 4 seconds) and the report only includes files from DSR (as expected).  

 ## What's next?
 - [This PR](https://github.com/vitest-dev/vitest/pull/3880) seems to implement a `--experimental-vm-threads` flag for Vitest to address issues with asset importing from external dependencies, but it does not yet seem to be available as of `v1.2.1`. 
- Seek alternative implementation methods for our [useIconSvg](https://github.com/cfpb/design-system-react/blob/main/src/hooks/useIconSvg.tsx) hook? 
- 🤷🏾‍♂️

**Comment by meissadia** [comment]
HMDA team uses [svgr](https://www.npmjs.com/package/vite-plugin-svgr) to convert their SVGs into React components.  I experimented with this approach and it works fine for our DSR icons. However we also want to integrate Design System's use of SVG (in plain JS) and adding this plugin seems to conflict with that goal (errors thrown).  I attempted to use the `exclude` configuration option to skip applying these tranformations to `@cfpb` libraries but was unsuccessful at bypassing the blocking errors

**Comment by meissadia** [comment]
I've also experimented with switching our coverage reporting library from `istanbul` to `v8`.  This change eliminates the excessive delay when generating the coverage report (v8 = ~5 seconds; istanbul = ~2 minutes).  Unfortunately, the report still includes the insane amount of unrelated files.
 
I have found that [setting an `include` pattern](https://github.com/istanbuljs/nyc?tab=readme-ov-file#using-include-and-exclude-arrays) while using `istanbul` restricts entries to files we *may* be interested in:
`vite.config.ts`
``` 
...
test: {
...
  coverage: {
    ...
    include: ['/src/*'],
  },
...
}
```

From the Istanbul docs on [Selecting files for coverage](https://github.com/istanbuljs/nyc?tab=readme-ov-file#selecting-files-for-coverage) , only code that is `require()`d should be included in the coverage report, so I don't know why we're seeing the results that we are.

**Comment by meissadia** [comment]
I think I've cracked it!  The issue is two-fold:
 
1. In test files, we're often importing from the DSR's centralized export (as in the code example below) causing ALL components in the library to be included in the coverage report.  By importing the components directly from their implementation files, we eliminate the unrelated components from the report. SO, I will open a separate ticket to eliminate that pattern from our test files. 
   ```
   // Bad for coverage reporting
   import { Alert, AlertFieldLevel } from '~/src/index';

   // Good for coverage reporting
   import { Alert } from './Alert.jsx'
   import { AlertFieldLevel } from './AlertFieldLevel.jsx'
   ```
   
2. Despite the [documentation for coverage reporter Istanbul](https://github.com/istanbuljs/nyc?tab=readme-ov-file#common-configuration-options) saying that the `all` option (which causes *all* code to be instrumented, measured, and included in the coverage report) defaults to `false`, I've found that explicitly setting it's value to `false` give us the expected coverage output, even without setting the `include` array of glob patterns to filter down the list.

**Comment by meissadia** [comment]
Fixes implemented!  Vitest tests passing.  Manual testing looks good. 
 
[Test deployment](https://deploy-preview-206--cfpb-design-system-react.netlify.app/?path=/docs/components-draft-select--overview)

**Review by natalia-fitzgerald — CHANGES_REQUESTED** [review]
@meissadia - As discussed in this morning's meeting, we will make some additional changes.

- Change component name from "Select" to "Selects"
- "Selects" overview should include DS intro for "Selects", link to source, "Types" (H4) and bullet list of links with "Single" and "Multiple"
- Sections should be "Single" and "Multiple"
- Section overviews should only include the link to the DS source page for Selects and no intro text
- Change first state for Single selects from "Default" to "Enabled"

**Comment by meissadia** [comment]
@natalia-fitzgerald Updated
@shindigira Many thanks for laying the ground work for this!

## Sidebar
Screenshot 2024-01-30 at 5 35 11 PM

## Overview
Screenshot 2024-01-30 at 5 35 19 PM

## Single
screencapture-localhost-6006-2024-01-30-17_36_09



## Multiple
Screenshot 2024-01-30 at 5 36 31 PM

**Review by natalia-fitzgerald — CHANGES_REQUESTED** [review]
@meissadia 
This is looking good. Just a couple of changes. 

Instead of "Default" for the label text in the examples can we uss "Label"? Or do you think it would be more parallel with Text inputs to remove the label entirely? 

In the DS we start with Option 1. What is Option 0?

**Comment by meissadia** [comment]
Updated labels to match the displayed element state:

Screenshot 2024-01-31 at 11 01 35 AM

**Comment by meissadia** [comment]
- Unit tests updated to match adjustments
- Screenshots in PR description updated to match latest state
- Ready for final review/merge!

**Review by natalia-fitzgerald — CHANGES_REQUESTED** [review]
The documentation and content look great. 

I noticed that the DSR tag is slightly taller than the DS tag. I spoke to @meissadia about this. I am not sure how we want to proceed - whether we want to push live (draft state) and then troubleshoot this (and fix in a later PR) OR wait until this is fixed. We can discuss.

File: src/components/Select/Select.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    disabled: { control: 'boolean' },
+    isMulti: { control: 'boolean' }
+  }
+};
+
+export default meta;
+
+type Story = StoryObj<typeof meta>;
+
+export const SingleSelect: Story = {
```
[the remark this answers is not in the document]
Comment (meissadia) [reply src/components/Select/Select.stories.tsx:19]: Great improvement! Thank you.
 
 
Screenshot 2024-02-01 at 1 58 35 PM

**Comment by meissadia** [comment]
> @meissadia The documentation and content look great!
> 
> I noticed that the DSR tag is slightly taller than the DS tag. I spoke to @meissadia about this. I am not sure how we want to proceed - whether we want to push live (draft state) and then troubleshoot this (and fix in a later PR) OR wait until this is fixed. We can discuss.
> 
> **CFPB Design System** Screenshot 2024-01-31 at 5 51 23 PM
> 
> **CFPB Design System in React (DSR)** Screenshot 2024-01-31 at 5 51 30 PM

@natalia-fitzgerald Since this issue is likely to be affecting other components that involve Icon + Text, I propose we move forward with merging the changes implemented in this PR and then addressing the icon alignment in a follow-up ticket. 
 
I've opened https://github.com/cfpb/design-system-react/issues/303 to track that investigation.

**Review by natalia-fitzgerald — APPROVED** [review]
Looks great @meissadia!

**Pull request merged**